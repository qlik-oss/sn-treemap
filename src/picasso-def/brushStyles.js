export const active = () => ({
  stroke: (d) => (d.type === 'text' ? d.stroke : 'black'),
  strokeWidth: (d) => {
    if (d.type === 'text') {
      return d.strokeWidth;
    }
    return d.data.child ? 0 : 2;
  },
  opacity: (d) => (d?.data?.overlay ? 0.7 : 1),
});

export const inactive = () => ({
  opacity: (d) => (d?.data?.parentGroup ? 0 : 0.5),
});
