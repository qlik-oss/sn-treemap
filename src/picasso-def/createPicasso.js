import Hammer from 'hammerjs'
import picassoHammer from 'picasso-plugin-hammer';
import picasso from 'picasso.js';
import picassoQ from 'picasso-plugin-q';

export const createPicasso = ({renderer}) => {
  const pic = picasso({
    renderer: {
      prio: [renderer || 'canvas'],
    },
  });
  pic.use(picassoQ);
  pic.use(picassoHammer);
  return {pic, picassoQ};
};
