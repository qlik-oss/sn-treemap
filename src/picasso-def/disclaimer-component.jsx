/** @jsx h */

function getStyle(dock, rect) {
  const style = {
    pointerEvents: 'auto',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    color: '#7b7a78',
    fontStyle: 'italic',
    fontSize: '13px',
  };

  switch (dock) {
    case 'left':
      style.width = `${rect.height}px`;
      style.transform = `rotate(-90deg) translate(-${rect.height}px, 0)`;
      style.transformOrigin = 'top left';
      break;

    case 'right':
      style.width = `${rect.height}px`;
      style.transform = `rotate(90deg) translate(0, -${rect.width}px)`;
      style.transformOrigin = 'top left';
      break;

    case 'center':
      style.width = `${rect.width}px`;
      style.height = `${rect.height}px`;
      style.color = '#595959';
      style.fontStyle = 'normal';
      style.fontSize = '1.2em';
      style.textAlign = 'center';
      break;

    default:
    case 'top':
    case 'bottom':
      style.justifyContent = 'flex-start';
      style.alignItems = 'flex-start';
      break;
  }
  return style;
}

export default {
  renderer: 'dom',
  disableTriggers: true,
  preferredSize: function preferredSize() {
    return 22;
  },
  render(h) {
    const {
      settings: { label, rtl },
      layout: { dock },
    } = this.settings;
    const dir = (dock === 'top' || dock === 'bottom') && rtl ? 'rtl' : 'ltr';
    const style = getStyle(dock, this.rect);

    const modifiedLabel = dock === 'center' ? label : `* ${label}`;

    const spanStyle = dock !== 'center' ? { overflow: 'hidden', whiteSpace: 'nowrap', textOverflow: 'ellipsis' } : {};

    return (
      <div style={style} dir={dir}>
        <span ref={this.alignToRef} style={spanStyle} title={modifiedLabel}>
          {modifiedLabel}
        </span>
      </div>
    );
  },
};
