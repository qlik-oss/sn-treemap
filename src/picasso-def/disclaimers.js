function blockingComponent(disclaimerLabel, translator, rtl) {
  return {
    key: 'blocking-disclaimer',
    type: 'disclaimer',
    layout: {
      displayOrder: 1000,
      dock: 'center',
    },
    settings: {
      label: translator.get(`Object.Disclaimer.${disclaimerLabel}`),
      rtl,
    },
  };
}

function noDataExist(layout) {
  if (!layout?.qHyperCube) {
    return true;
  }
  const { qcx, qcy } = layout.qHyperCube.qSize;
  return qcx * qcy === 0;
}

function onlyNanData(layout) {
  return layout.qHyperCube.qMeasureInfo[0]?.qMin === 'NaN' && layout.qHyperCube.qMeasureInfo[0]?.qMax === 'NaN';
}

function onlyNegativeOrZeroValues(layout) {
  return layout.qHyperCube.qMeasureInfo[0]?.qMax <= 0;
}

export function getBlockingDisclaimer(layout, translator, rtl) {
  if (noDataExist(layout)) {
    return blockingComponent('NoDataExist', translator, rtl);
  }
  if (onlyNanData(layout)) {
    return blockingComponent('OnlyNanData', translator, rtl);
  }
  if (onlyNegativeOrZeroValues(layout)) {
    return blockingComponent('OnlyNegativeOrZeroValues', translator, rtl);
  }
  return null;
}

function negativeOrZeroValues(layout) {
  return layout.qHyperCube.qMeasureInfo[0]?.qMin <= 0;
}

function infoComponent(disclaimerLabel, translator, rtl) {
  return {
    key: 'info-disclaimer',
    type: 'disclaimer',
    layout: {
      displayOrder: 1000,
      dock: 'bottom',
    },
    settings: {
      label: translator.get(`Object.Disclaimer.${disclaimerLabel}`),
      rtl,
    },
  };
}

export function getInfoDisclaimer(layout, translator, rtl) {
  if (layout.showDisclaimer === false) {
    return undefined;
  }
  if (negativeOrZeroValues(layout)) {
    return infoComponent('NegativeOrZeroValues', translator, rtl);
  }
  return undefined;
}
