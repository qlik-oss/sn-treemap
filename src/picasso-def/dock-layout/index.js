import { LAYOUT_MODES } from './layout-modes';
import { logicalSize } from './logica-size';

export function dockLayout(layout, options) {
  return {
    layoutModes: LAYOUT_MODES,
    center: {
      minWidthRatio: 0,
      minHeightRatio: 0,
    },
    logicalSize: logicalSize(layout, options),
  };
}
