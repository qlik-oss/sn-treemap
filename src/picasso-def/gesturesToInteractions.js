export default function gesturesToInteractions(interactionType, gestures) {
  return {
    type: interactionType,
    key: 'hammer',
    gestures,
  };
}
