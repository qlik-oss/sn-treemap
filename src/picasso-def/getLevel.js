const isSelectableDim = (info, index) => {
  if (info.qLocked) {
    return false;
  }
  if (info.qStateCounts.qSelected > 1) {
    return true;
  }
  if (info.qStateCounts.qOption > 0) {
    return true;
  }
  if (info.qStateCounts.qSelected === 1 && info.qStateCounts.qAlternative > 0) {
    return info.qCardinalities.qHypercubeCardinal > 1 + index; // + index is for total nodes
  }
  return false;
};

export const getNextSelectLevel = ({ qHyperCube }) => {
  // find the first dimension that is selectable:
  // - the dimension must not be a pseudo dimension
  // - the dimension must not be locked
  // - if the dimension has been selected in, there must be other nodes to select in that same dimension

  const { qDimensionInfo } = qHyperCube;
  return qDimensionInfo.findIndex(isSelectableDim);
};

export const getAutoColorLevel = ({ qHyperCube }) => {
  const { qDimensionInfo } = qHyperCube;
  const dim = qDimensionInfo.findIndex((info, index) => info.qCardinalities.qHypercubeCardinal > 1 + index);
  return dim === -1 ? qDimensionInfo.length - 1 : dim;
};
