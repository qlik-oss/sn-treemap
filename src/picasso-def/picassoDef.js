import customTooltipNodes from '../custom-tooltip/picasso-definitions/nodes';
import { getLevel, getNextSelecteDim } from './getLevel';
import ChartFormatting from './formatting/chart-formatting';
import { legend } from './legend';
import createSelectables from './selectables';
import createTooltipService from './tooltip/service';
import { active, inactive } from './brushStyles';
import { dockLayout } from './dock-layout';
import gesturesToInteractions from './gesturesToInteractions';
import { getBlockingDisclaimer, getInfoDisclaimer } from './disclaimers';

const getFormatterForMeasures = (localeInfo, nrMeasures, qMeasureInfo) => {
  let measure;
  const chartFormatter = new ChartFormatting(localeInfo, qMeasureInfo);
  let f;
  let formatters = [];

  for (let i = 0; i < nrMeasures; i++) {
    measure = qMeasureInfo[i];
    f = chartFormatter.getMeasureFormatter(i);
    if (f && f.type === 'U') {
      f.pattern = f.createPatternFromRange(measure.qMax, measure.qMax, true);
      if (Math.floor(Math.log(Math.abs(measure.qMax)) / Math.log(10)) % 3 === 2) {
        f.pattern = f.pattern.substr(0, f.pattern.length - 2);
        f.pattern += 'A';
      }
      f.prepare();
    }

    formatters.push(f);
  }

  // the api importing returns  undefined formatters, filter them out, then
  // add a default one if necessary.
  formatters = formatters.filter((formatter) => formatter !== undefined);

  if (formatters.length === 0) {
    // add default formatter
    formatters.push({
      formatValue: (v) => v,
    });
  }
  return formatters;
};

export const picassoDef = ({
  layout,
  theme,
  env,
  showLegend,
  invalidMessage,
  translator,
  colorService,
  chart,
  options,
  actions,
  customTooltipService,
  properties,
}) => {
  const blockingDisclaimer = getBlockingDisclaimer(layout, translator);
  if (blockingDisclaimer) {
    return {
      components: [blockingDisclaimer],
      strategy: dockLayout(layout, options),
    };
  }

  if (!layout.qHyperCube) {
    return {};
  }
  const level = getLevel(layout);
  const dimLevel = getNextSelecteDim(layout);
  const selectLevel = Math.min(level, dimLevel);
  const interactionType = env.carbon ? 'kinesics' : 'hammer';
  const { qMeasureInfo } = layout.qHyperCube;

  const formatter = getFormatterForMeasures('', qMeasureInfo.length, qMeasureInfo);

  const scales = colorService.getScales();

  const treemapLegend = legend({ colorService, chart, layout });
  const selectables = createSelectables({
    actions,
    colorService,
    isDimensionLocked: false,
    isSingleSelectionmodels: false,
    scales,
    legendComponent: treemapLegend?.components[0],
  });
  const gestures = selectables.gestures.sort((a, b) => (b.prio || 0) - (a.prio || 0));

  const tooltipService = createTooltipService({
    chart,
    actions,
    translator,
    rtl: false, // TODO
    colorService,
    theme,
    custom: customTooltipService,
    properties,
    level,
    layout,
    formatter,
  });

  const dimensionCount = layout.qHyperCube.qDimensionInfo.length;
  const collections = [
    {
      key: 'hierarchy',
      data: {
        hierarchy: {
          label(datum) {
            return datum?.qText ? datum.qText : undefined;
          },
          value(datum) {
            if (datum?.qType === 'N' && datum?.qSubNodes?.length > 1) {
              return 'NaN';
            }
            if (datum?.qType === 'N' && datum?.qSubNodes?.length === 1) {
              let sb = datum.qSubNodes;
              while (sb.length > 1) {
                sb = sb.qSubNodes;
              }
              if (sb[0].qType !== 'V') {
                return 0;
              }
              return datum.qMaxPos;
            }
            if (datum.qType === 'O' && datum?.qSubNodes?.length === 1) {
              return datum?.qSubNodes[0].qValue || 0;
            }

            return 0;
          },
          props: {
            select: {
              field: `qDimensionInfo/${selectLevel}`,
              reduce: (values) => (values.length === 1 ? values[0] : undefined),
            },
            ...colorService.getDatumProps(),
            isOther: { value: (d, node) => node.data.qElemNo === -3 },
            customTooltipAttrExps:
              dimensionCount !== 0 ? customTooltipNodes.getNode(layout, { dimensionCount }) : undefined,
          },
        },
      },
    },
  ];

  const brushSettings = {
    consume: [
      {
        filter: (d) => d?.data?.select?.value > -1,
        context: 'selection',
        data: ['select', 'fill'],
        style: {
          active: active(),
          inactive: inactive(),
        },
      },
    ],
  };

  const components = [
    {
      type: 'treemap',
      key: 'treemap',
      data: { collection: { key: 'hierarchy' } },
      settings: {
        formatter,
        labels: layout.labels,
        box: {
          fill: colorService.getColor(),
        },
        color: {
          mode: layout.color.mode,
          measureScheme: layout.color.measureScheme,
          others: theme.getDataColorSpecials().others,
          persistent: layout.color.persistent || layout.qHyperCube.qDimensionInfo.length > 1,
        },
        showHeaders: layout.qHyperCube.qDimensionInfo.length > 1,
        headerColor: '#F2F2F2',
        level,
        invalidMessage,
        translator,
        theme,
      },
      brush: brushSettings,
    },
  ];

  if (treemapLegend && showLegend === true) {
    components.push(...treemapLegend.components);
    gestures.unshift(...treemapLegend.interactions);
  }

  components.push(...tooltipService.getComponents());
  const tooltipInteractions = tooltipService.getInteractions();
  gestures.push(...tooltipInteractions.gestures);

  components.push(...selectables.components);

  const infoDiscalimer = getInfoDisclaimer(layout, translator);
  if (infoDiscalimer) {
    components.push(infoDiscalimer);
  }

  const interactions = [tooltipInteractions.native, gesturesToInteractions(interactionType, gestures)];

  return {
    collections,
    scales,
    components,
    interactions,
    palettes: colorService.getPalettes(),
    strategy: dockLayout(layout, options),
  };
};
