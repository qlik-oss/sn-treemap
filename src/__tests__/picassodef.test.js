/* eslint-disable no-underscore-dangle */

import { __DO_NOT_USE__ as NebualInternals } from '@nebula.js/stardust';
import { picassoDef } from '../picasso-def/picassoDef';
import testdata from './testdata.json';

describe('Treemap picassodef render', () => {
  const { theme: themeFn } = NebualInternals;
  it('should render ok', () => {
    const layout = testdata.qLayout[0].value;
    const _theme = themeFn();
    const colorService = {
      getDatumProps: () => ({
        color: {
          field: 'qMeasureInfo/0',
        },
        expressionColor: undefined,
        expressionColorText: undefined,
      }),
      getScales: () => ({
        color: {
          data: {
            extract: { field: 'qMeasureInfo/0' },
          },
          range: ['#ec983d', '#deb357', '#d0ce71', '#c3ea8c', '#8ed195', '#5ab89e', '#26a0a7'],
          type: 'color',
        },
      }),
      getColor: () => undefined,
      getLegend: () => undefined,
      getPalettes: () => undefined,
      getSettings: () => ({}),
    };
    const data = picassoDef({ layout, theme: _theme.externalAPI, env: {}, colorService });
    expect(data).toMatchSnapshot();
  });
});
