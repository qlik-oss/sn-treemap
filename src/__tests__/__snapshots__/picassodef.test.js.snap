// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Treemap picassodef render should render ok 1`] = `
{
  "collections": [
    {
      "data": {
        "hierarchy": {
          "children": [Function],
          "label": [Function],
          "props": {
            "color": {
              "field": "qMeasureInfo/0",
            },
            "customTooltipAttrExps": {
              "field": "qDimensionInfo/0",
              "value": [Function],
            },
            "expressionColor": undefined,
            "expressionColorText": undefined,
            "isNotFetchedOthers": {
              "value": [Function],
            },
            "isNull": {
              "value": [Function],
            },
            "isOther": {
              "value": [Function],
            },
            "select": {
              "field": "qDimensionInfo/0",
              "reduce": [Function],
            },
            "size": {
              "label": [Function],
              "value": [Function],
            },
          },
          "value": [Function],
        },
      },
      "key": "hierarchy",
    },
  ],
  "components": [
    {
      "brush": {
        "consume": [
          {
            "context": "selection",
            "data": [
              "select",
              "fill",
            ],
            "style": {
              "active": {
                "opacity": [Function],
                "stroke": [Function],
                "strokeWidth": [Function],
              },
              "inactive": {
                "opacity": [Function],
              },
            },
          },
        ],
      },
      "data": {
        "collection": {
          "key": "hierarchy",
        },
      },
      "key": "treemap",
      "settings": {
        "box": {
          "fill": undefined,
        },
        "color": {
          "measureScheme": "dc",
          "mode": "byMeasure",
          "others": "#a5a5a5",
          "persistent": false,
        },
        "headerColor": "#F2F2F2",
        "invalidMessage": undefined,
        "labels": {
          "auto": true,
          "headers": true,
          "leaves": true,
          "overlay": true,
          "values": true,
        },
        "rtl": false,
        "selectLevel": 0,
        "showHeaders": false,
        "theme": {
          "_events": {},
          "_maxListeners": undefined,
          "addListener": [Function],
          "emit": [Function],
          "getColorPickerColor": [Function],
          "getContrastingColorTo": [Function],
          "getDataColorPalettes": [Function],
          "getDataColorPickerPalettes": [Function],
          "getDataColorScales": [Function],
          "getDataColorSpecials": [Function],
          "getStyle": [Function],
          "listeners": [Function],
          "name": [Function],
          "on": [Function],
          "once": [Function],
          "removeAllListeners": [Function],
          "removeListener": [Function],
          "setMaxListeners": [Function],
        },
        "translator": undefined,
      },
      "type": "treemap",
    },
    {
      "beforeMount": [Function],
      "beforeUpdate": [Function],
      "destroyed": [Function],
      "key": "tooltip-qcm",
      "layout": {
        "displayOrder": 3,
      },
      "settings": {
        "afterHide": [Function],
        "afterShow": [Function],
        "appendTo": [Function],
        "content": [Function],
        "direction": "ltr",
        "duration": 8000,
        "extract": [Function],
        "filter": [Function],
        "placement": [Function],
      },
      "show": true,
      "style": {
        "arrow": {
          "color": "#404040",
        },
        "cell": {
          "hyphens": "auto",
          "max-width": "180px",
          "overflow-wrap": "break-word",
          "word-break": "break-word",
          "word-wrap": "break-word",
        },
        "content": {
          "background": "rgba(64, 64, 64, 0.9)",
          "border-spacing": "4px",
          "display": "table",
          "empty-cells": "show",
          "fontFamily": "'Source Sans Pro', 'Arial', 'sans-serif'",
          "fontSize": "13px",
          "opacity": "1",
        },
      },
      "type": "tooltip",
    },
    {
      "key": "lasso",
      "layout": {
        "displayOrder": 99,
      },
      "settings": {
        "brush": {
          "components": [
            {
              "action": "add",
              "contexts": [
                "selection",
              ],
              "data": [
                "select",
              ],
              "key": "treemap",
            },
          ],
        },
        "lasso": {
          "fill": "rgba(0, 0, 0, 0)",
          "opacity": 0.7,
          "stroke": "black",
          "strokeDasharray": [],
          "strokeWidth": 2,
        },
        "snapIndicator": {
          "opacity": 0.7,
          "stroke": "black",
          "strokeDasharray": "5, 5",
          "strokeWidth": 2,
          "threshold": 75,
        },
        "startPoint": {
          "opacity": 1,
          "r": 10,
          "stroke": "rgba(77, 192, 78, 1)",
          "strokeWidth": 0,
        },
      },
      "type": "brush-lasso",
    },
  ],
  "interactions": [
    {
      "events": {
        "mouseleave": [Function],
        "mousemove": [Function],
        "mouseup": [Function],
        "wheel": [Function],
      },
      "type": "native",
    },
    {
      "gestures": [
        {
          "events": {
            "tap": [Function],
          },
          "key": "event:tap",
          "options": {
            "enable": [Function],
            "event": "tap",
            "interval": 10,
            "pointers": 1,
          },
          "recognizeWith": undefined,
          "requireFailure": undefined,
          "type": "Tap",
        },
        {
          "events": {
            "lassoend": [Function],
            "lassomove": [Function],
            "lassostart": [Function],
          },
          "key": "event:lasso",
          "options": {
            "enable": [Function],
            "event": "lasso",
          },
          "recognizeWith": undefined,
          "requireFailure": undefined,
          "type": "Pan",
        },
        {
          "events": {
            "tooltipTap": [Function],
          },
          "key": "event:tooltipTap",
          "options": {
            "enable": [Function],
            "event": "tooltipTap",
          },
          "recognizeWith": "tap",
          "type": "Tap",
        },
      ],
      "key": "hammer",
      "type": "hammer",
    },
  ],
  "palettes": undefined,
  "scales": {
    "color": {
      "data": {
        "extract": {
          "field": "qMeasureInfo/0",
        },
      },
      "range": [
        "#ec983d",
        "#deb357",
        "#d0ce71",
        "#c3ea8c",
        "#8ed195",
        "#5ab89e",
        "#26a0a7",
      ],
      "type": "color",
    },
  },
  "strategy": {
    "center": {
      "minHeightRatio": 0,
      "minWidthRatio": 0,
    },
    "layoutModes": {
      "FULL": {
        "height": 450,
        "width": 550,
      },
      "MEDIUM": {
        "height": 220,
        "width": 400,
      },
      "SMALL": {
        "height": 150,
        "width": 300,
      },
      "SPARK": {
        "height": 0,
        "width": 0,
      },
      "XSMALL": {
        "height": 75,
        "width": 150,
      },
    },
    "logicalSize": undefined,
  },
}
`;
